beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Instant;
+
+/** A {@link PTransform} that reads from a {@link CommitLogReader} using splittable DoFn. */
+public class CommitLogRead extends PTransform<PBegin, PCollection<StreamElement>> {
+
+  /**
+   * Create the {@link CommitLogRead} transform.
+   *
+   * @param observeName name of the observer
+   * @param position {@link Position} to read from
+   * @param limit limit (use {@link Long#MAX_VALUE} for unbounded
+   * @param repo repository
+   * @param reader the reader
+   * @return {@link CommitLogRead} transform for the commit log
+   */
+  public static CommitLogRead of(
+      String observeName, Position position, long limit, Repository repo, CommitLogReader reader) {
+
+    return new CommitLogRead(observeName, position, limit, repo, reader);
+  }
+
+  @DoFn.UnboundedPerElement
+  private static class CommitLogReadFn extends DoFn<byte[], StreamElement> {
+
+    private final String name;
+    private final Position position;
+    private final long limit;
+    private final RepositoryFactory repositoryFactory;
+    private final CommitLogReader.Factory<?> readerFactory;
+
+    private transient boolean externalizableOffsets = false;
+    private transient BlockingQueueLogObserver observer;
+
+    private CommitLogReadFn(
+        String name,
+        Position position,
+        long limit,
+        RepositoryFactory repositoryFactory,
+        CommitLogReader.Factory<?> readerFactory) {
+
+      this.name = name;
+      this.position = position;
+      this.limit = limit;
+      this.repositoryFactory = repositoryFactory;
+      this.readerFactory = readerFactory;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(
+        RestrictionTracker<OffsetRange, Offset> tracker, OutputReceiver<StreamElement> output) {
+
+      try (ObserveHandle handle = startObserve(tracker.currentRestriction())) {
+        int i = 0;
+        while (!Thread.currentThread().isInterrupted()
+            && observer.getWatermark() < Watermarks.MAX_WATERMARK) {
+
+          StreamElement element = observer.take();
+          if (element == null) {
+            break;
+          }
+          boolean isFirst = i++ == 0;
+          Offset offset = Objects.requireNonNull(observer.getLastReadContext().getOffset());
+          if (externalizableOffsets
+              && !tracker.currentRestriction().isStartInclusive()
+              && isFirst) {
+            // skip the exclusive first offset
+            continue;
+          }
+          if (!tracker.tryClaim(offset)) {
+            return ProcessContinuation.stop();
+          }
+          output.outputWithTimestamp(element, Instant.ofEpochMilli(element.getStamp()));
+        }
+        observer.stop();
```

Comment:
this should be moved to `finally` block, as we can return earlier (L112)

can we make a failing test case for this?

Reply:
I think that the call to `observer.stop()` is actually redundant. Definitely not worth adding a test case, if anything, we can remove it completely. :-)

Reply:
Hm, it is be even actually wrong to call it here. This PR is far from complete, so the fact, that it is wrong will be unrevealed when testing a commit-log with non-externalizable offsets (like PubSub). I'll remove the call completely.

Reply:
I was wrong, it was necessary due to how we fill the blocking queue. A refactored it a bit to do just what is needed and put it to the finally block as suggested. :+1: